MIXEDARCH-80: Guide on using node affinity to safely schedule pods in heterogeneous clusters. (#5983)
operator-framework/operator-sdk: website/content/en/docs/advanced-topics/heterogeneous.md

---
title: Heterogeneous Cluster Support
linkTitle: Heterogeneous Cluster Support
weight: 80
---
## Overview
A heterogeneous cluster is a cluster in which the nodes differ from each other in a way that makes their containers incompatible with each other. The most common incompatibility that applies is CPU architecture. The rest of this document assumes that the user is developing an application that needs to be able to deploy cleanly in a cluster with a mix of node CPU architectures. Because an operator deployed to a heterogeneous cluster may not support all of the architectures in the cluster, it is important to set up safeguards that prevent scheduling pods on nodes of incompatible architectures. These scheduling implications affect both the operator as well as its operands.

> ### Determining the Supported Architectures for an Operator/Operand Image

Reviewer: ```suggestion
### Determining the Supported Architectures for an workloads images
```

For a given container image, you can check which platforms are supported by listing them by inspecting the manifest.
```bash
docker manifest inspect <image>
```

> If the above command fails, it may be because the referenced name points to a single image rather than a manifest list. In this case, you can find the architecture by inspecting the image.

Reviewer: Not only the arch matters right? So, I think this explanation can lead to a mistake.
If you follow this one you might have an image that supports darwin/arch64 and you will configure it to work in a node that is linux/arm64. 

c/c @jmrodri

Author: So while OS matters, its much less relevant when we talk about heterogeneous clusters. It would be a very strange set-up that called for some nodes on Windows and some on Linux for example.

You mention darwin as a possibility, but that would generally have to involve clusters that span across a workstation and some external source of nodes. It's not that these are impossible, but they're very unlikely.

--

That said, I we rework the working so that it's a little less "architecture" and a little more "platform".

Reviewer: Hi @jaypoulz , 

My point is that the image can have support **for example** to SO/arm64 and NOT linux/arm64
Then, following this guide I will be configuring it to work in nodes with Linux/arm54 which that could fail.

```bash
docker inspect <image>
```

> Here's an example of the architectures distributed for the Alpine Linux container image.

Reviewer: WDYT about add the commend as example which contains the image and the whole result with arch and os which lead a better understanding

Author: Do you mean the full output for both commands? (I.e. docker inspect & docker manifest inspect)

I felt like the output for Alpine was too long, which is why I truncated it via the pipe to grep. I can provide a truncated version of the full output if that's preferable.

Reviewer 2: I agree with @jaypoulz, I'd much rather see the limited list of items that are relevant than a bunch of noise. I can run the command myself to see that.

Reviewer: @jaypoulz see: https://github.com/operator-framework/operator-sdk/pull/5983/files#r967708843
You do not need to add all just copy and paste 2 so and archs. 

Example:

```json
...
{
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 528,
         "digest": "sha256:fa30af02cc8c339dd7ffecb0703cd4a3db175e56875c413464c5ba46821253a8",
         "platform": {
            "architecture": "ppc64le",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 528,
         "digest": "sha256:c2046a6c3d2db4f75bfb8062607cc8ff47896f2d683b7f18fe6b6cf368af3c60",
         "platform": {
            "architecture": "s390x",
            "os": "linux"
         }
      }

...
```

```bash
$ docker manifest inspect alpine | grep architecture
            "architecture": "amd64",
            "architecture": "arm",
            "architecture": "arm",
            "architecture": "arm64",
            "architecture": "386",
            "architecture": "ppc64le",
            "architecture": "s390x",
```

#### Shipping a Multi-Architecture Operator
If once you've determined which subset of architecture nodes you want to distribute your operator for, there are tools available to help you distribute your operator.

> OperatorSDK provides a validator to help check bundle configuration.

Reviewer 2: Operator SDK should be written as 2 words or as the command `operator-sdk`.

```suggestion
Operator SDK provides a validator to help check bundle configuration.
```

> OperatorSDK provides a validator to help check bundle configuration.

Reviewer: heterogeneous and multiple arch is the same thing right?
It seems like it is not by reading the doc

IHMO speaking about the validator here is confusing because I think that we should explain how things work and what are the pre-requirements and recommendations. After all, we can say something like:

 if you are integrating your operator with OLM then you will create a bundle, the bundle will be created based on the configuration provided in the config/ directory. To help you out validate the multiple architecture configuration you can try to test your bundle with operator-sdk bundle validate ./bundle --select-optional name=multiarch.

However, note that all workloads created in your project must be configured accordingly as well and the above validator is capable of checking only what is configured in the bundle CSV manifest.

Author: Heterogeneous and multi-arch do not typically mean the same. Thing thing.

All clusters that have nodes of different architectures are heterogeneous.
Not all heterogeneous clusters involve nodes of different architectures. For example, you mentioned "darwin/arm64" vs "linux/arm64". Theoretically, you could have a homogeneous architecture cluster that is heterogeneous by operating system.

Finally, when we talk about multi-arch (as a general rule), we are usually talking about homogeneous clusters on a architecture other than amd64. That is - all nodes are uni-formally arm64, ppc64le, or s390x. It is only recently that we've extended the multi-architecture scope to include non-homogeneous environments.

--

With specific regard to the validator - I was intending on following up this patch with either:
1. a heterogeneous validator
2. an extension to the multi-arch validator to include heterogeneous considerations

So while the discussion is incomplete, I do think it's important to highlight was is already available to help start this journey. To be clear, talking about heterogeneous support generally isn't relevant until the cluster already support multiple architectures.

```bash
operator-sdk bundle validate ./bundle --select-optional name=multiarch
```

> If you're planning to distribute the operator via OLM, you can find more information in the [OLM guide for supporting multiarch](https://olm.operatorframework.io/docs/advanced-tasks/ship-operator-supporting-multiarch/).

Reviewer: Why someone would use the validator if is not planned to distribute in OLM?
That bundle is only created for you to integrate the project with OLM and distribute with.

Author: I have never used the validator myself, so I was unsure of exactly what it does. I will adjust this section to make it clear that the validator fits under the umbrella of things you might use when integrating w/ OLM.

> ### Setting the Supported Architectures Operator & Operand Workloads

Reviewer 2: ```suggestion
### Setting the Supported Architectures on the Operator & Operand Workloads
```
This heading feels like it needs something. Either "on the Operator & ..." or "for the Operator & ..."

[Kubernetes provides a mechanism](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/) called `nodeAffinity` which can be used to limit the possible node targets where a pod can be scheduled. The following example can be used to update a `PodSpec` or `PodTemplateSpec` to prevent the scheduling of pods on nodes of incompatible architectures. Here we compare the `kubernetes.io/arch` key set on the node to ensure that the value matches one of the supported architectures. This assumes that the referenced image points to a manifest list of images for each of the architectures supported. It is important to remember that `nodeAffinity` must be set anywhere a container reference is defined, including `Pod`, `Deployment`, `DaemonSet`, `StatefulSet`, or any other object that defines a `PodSpec` or `PodTemplateSpec`.

```yaml
affinity:
  nodeAffinity:
    requiredDuringSchedulingIgnoredDuringExecution:
      nodeSelectorTerms:
      - matchExpressions:
        - key: kubernetes.io/arch
          operator: In
          values:
          - amd64
          - arm64
          - ppc64le
          - s390x

> ```

Reviewer 3: can we also add the "operating system" piece in the `matchExpressions` in the code snippet too?

We've started to get asks around how to specify the workloads managed by Operators to be scheduled to the linux node pool in a Windows container/Windows node enabled mixed cluster. Hence it'll be good to have a single doc that covers how to specify arch and OS in nodeAffinity so folks can follow.

- matchExpressions:
          - key: kubernetes.io/os
            operator: In
            values:
              - linux

Reviewer 4: @tlwu2013 pretty sure that example boils down to `(arch == amd64 || arch == arm64 || arch == ppc64le || arch == s390x) || (os == linux)`, which doesn't seem like what you want?

Seems like you'd actually want:
```yaml
affinity:
  nodeAffinity:
    requiredDuringSchedulingIgnoredDuringExecution:
      nodeSelectorTerms:
      - matchExpressions:
        - key: kubernetes.io/arch
          operator: In
          values:
          - amd64
          - arm64
          - ppc64le
          - s390x
        - key: kubernetes.io/os
          operator: In
          values:
            - linux   
```

Reviewer 4: I am also worried about [OLM's CSV multi-arch annotations](https://olm.operatorframework.io/docs/advanced-tasks/ship-operator-supporting-multiarch/#supported-labels) which are ambiguous if multiple architectures and OSes are supported.

If someone had an operator that supported windows/amd64 and linux/arm64, there's no way to precisely assert that with OLM's annotations. You'd probably end up putting:
```
labels:
    operatorframework.io/arch.amd64: supported
    operatorframework.io/arch.arm64: supported
    operatorframework.io/os.linux: supported
    operatorframework.io/os.windows: supported
```

Reviewer 4: Seems like we need to deprecate those labels and replace them with something like:
```yaml
labels:
    operatorframework.io/platform.windows_amd64: supported
    operatorframework.io/platform.linux_arm64: supported
```

Reviewer 3: > @tlwu2013 pretty sure that example boils down to `(arch == amd64 || arch == arm64 || arch == ppc64le || arch == s390x) || (os == linux)`, which doesn't seem like what you want?
> 
> Seems like you'd actually want:
> 
> ```yaml
> affinity:
>   nodeAffinity:
>     requiredDuringSchedulingIgnoredDuringExecution:
>       nodeSelectorTerms:
>       - matchExpressions:
>         - key: kubernetes.io/arch
>           operator: In
>           values:
>           - amd64
>           - arm64
>           - ppc64le
>           - s390x
>         - key: kubernetes.io/os
>           operator: In
>           values:
>             - linux   
> ```

good catch @joelanford, the example I pasted was copied from https://github.com/kubernetes-sigs/kubebuilder/pull/2906. So sounds like that might need to be changed too.

The list of architecture values should only include the architectures the operator supports. The full list of possible `GOARCH` values is [available here](https://github.com/golang/go/blob/master/src/go/build/syslist.go). It should be noted that Kubernetes allows a user to specify both `requiredDuringSchedulingIgnoredDuringExecution` and `preferredDuringSchedulingIgnoredDuringExecution`. The syntax for each block is the same. At a minimum, an operator author should set the required terms since they protect against a pod being scheduled on an architecturally incompatible node. If the operator in question has much better performance on a subset of the supported architectures, it may also be prudent to set `preferredDuringSchedulingIgnoredDuringExecution` so that the optimal arches are selected first if they are available.

#### Integrating with Operator Lifecycle Manager

> The Operator Lifecycle Manager (OLM) is often used to distribute operators via operator catalogs. The operator deployment object in an OLM integrated operator is defined in the `ClusterServiceVersion` yaml. It is important to remember to set the node affinity block in each of the `spec.template.spec` under `spec.install.deployments`.

Reviewer 4: The CSV's `spec.install.deployments` block is automatically generated from the operator deployment generated by `kustomize build ./config`. So the correct place to add the node affinity stuff is in the `./config/manager/manager.yaml` file.

I wonder if we could separate the upstream, non-OLM concepts (e.g. manifest lists, setting node affinity, etc) from the OLM-specific bits (setting the CSV supported os/arch annotations)

Reviewer 4: Follow up to previous comment: For those _not_ using OLM, the doc should have a clear set of things to do to ensure the operator/operands are setup with the correct manifest list images and node affinity configs. And that set of stuff has nothing to do with bundles or CSVs.

Then in this `Integrating with OLM` section, we could talk about what _extra_ stuff someone should do to build/validate a heterogeneous-compatible bundle.

Reviewer 4: By the way, I just want to call to your attention that there's a whole separate docs section for integrating with OLM here: https://sdk.operatorframework.io/docs/olm-integration/

I'm not sure if the docs site already has most/all of the OLM-related stuff under that section or whether we've let OLM references spread throughout all the sections.

Either way, it might be good to take note of how other docs bring in the OLM integrations and do something similar for this one.

## Examples
This section attempts to show how each of the sample operators could be updated to declare a set of supported architectures. Most of the time, these changes will be in `yaml` files, but different kinds of operators embed `PodSpec` and `PodTemplateSpec` objects in different places.

> 

Reviewer: In all projects we have the same config by default, see: https://sdk.operatorframework.io/docs/overview/project-layout/.

The only places that we need to configure the node affinity are for the deployment used to run the manager and the sidecar proxy. (no matter the language)

Then, note that for the proxy we can/will set the nodeAffinity by default. See: https://github.com/kubernetes-sigs/kubebuilder/pull/2906/files#diff-cd538f6a8d1b3f24837737051cb800f6ab335924d171ba78db29b8690cd55b46
(so users will not need to update this one)

For the manager, we cannot set it because that depends on how they build the image. But we will add the code commented to try encourage and help them out see: https://github.com/kubernetes-sigs/kubebuilder/pull/2906/files#diff-72ce7e8fc292f0c680218ed1cfce019d7baa9c0e41218b8526eecfd57d0bf344

Also, in the same PR, we are adding a new target to help them out to build the images using buildx.

> The most common directory with objects that will need an architecture affinity block is the operator's `config` directory. Additionally, if the operator is configured for OLM, there will likely be a deployment object in the `bundle/manifests/*.clusterserviceversion.yaml`. These locations are common across all operator types.

Reviewer: That is not the most common. 
The only directory scaffold by default that has manifest is the config/

Ansible/Helm operators will have other directories which are the Ansible playbooks and helm charts. 
So, we might want to clarify that any workload created by the project would need to have the config, for Golang they must check/ensure in the controller/reconciliation code. For Ansible/Helm they will need to ensure that the playbooks and charts are configured accordingly.

Author: Is the confusion here is less that `config` is the **most common** and rather that it's the directory that all operator types have **in common** where updates maybe be needed?

Reviewer: @jaypoulz the projects use kustomize so the patches ones only add things in the Deployment. However, we have only 1 Deployment with 2 containers/images by default.

> Each of these examples can be in the [OperatorSDK GitHub repository](https://github.com/operator-framework/operator-sdk).

Reviewer 2: ```suggestion
Each of these examples can be in the [Operator SDK GitHub repository](https://github.com/operator-framework/operator-sdk).
```

### Ansible Operators
In the `testdata/ansible` samples app, you'll find several configuration objects that need a nodeAffinity specification in the `config` and `bundle/manifests` directories. The full list is as follows:
- `testdata/ansible/memcached-operator/bundle/manifests/memcached-operator.clusterserviceversion.yaml`
- `testdata/ansible/memcached-operator/config/default/manager_config_patch.yaml`
- `testdata/ansible/memcached-operator/config/default/manager_auth_proxy_patch.yaml`

> - `testdata/ansible/memcached-operator/config/manager/manager.yaml`

Reviewer: I do not think that we should use testdata/ansible. It does not exist when users create their projects with the tool.
The testdata is a dir that only exists in the SDK repo. Usually, we describe how things work in the project generated by the tool and we link the samples/examples or we say you can check it in the testdata/sample. 

However, provide the path `testdata/ansible/` is strange. 

Also, note that only the manager and manager_auth_proxy_patch.go (which will patch the manager to add the auth proxy container) are those that should be changed.

Author: I can certainly simplify the path.
One question I had when starting this is whether it would be more helpful to actually link the content. My worry is that if we link the content, it may change with updates over time that would break the links.

Can you explain to me what the `patch.yaml` file does so I can understand why it doesn't need the update?

Reviewer: the patches do not define the container
Then, patch things such as add an envvar or an extra config.
See how to config and where the nodeAffinity by looking at the changes  proposed in: https://github.com/kubernetes-sigs/kubebuilder/pull/2906

The Ansible example also has Pod-defining objects as part of the test directories. For completeness, these should also need to define a nodeAffinity specification. The operator author will know best if tests are expected to be run in mixed-architecture environments, but updating these helps future-proof the operator.
- `testdata/ansible/memcached-operator/config/testing/debug_logs_patch.yaml`
- `testdata/ansible/memcached-operator/config/testing/manager_image.yaml`
- `testdata/ansible/memcached-operator/config/testing/pull_policy/Always.yaml`
- `testdata/ansible/memcached-operator/config/testing/pull_policy/IfNotPresent.yaml`

> - `testdata/ansible/memcached-operator/config/testing/pull_policy/Never.yaml`

Reviewer: The only Deployment define here is in the manager_image (but that is for test the project only)
Note that the other files are patches to add envars and etcs. 

Ansible/Helm also uses the same structure for Golang and the projects by default scaffolds one 2 workloads which are the containers for the manager and the proxy. 

Check the PR for the docs in KB: https://github.com/kubernetes-sigs/kubebuilder/pull/2906/files#diff-b79471388fbaeb59d4a9793846e4b1c0e533f34c4768b5fad4907338b90c74a8

Btw, I think we can link the KB doc and just add here what is SDK specific. 

c/c @jmrodri

 Ansible operators may also create Pod-defining objects dynamically. An example of this kind of object would be in `testdata/ansible/memcached-operator/roles/memcached/tasks/main.yml`.
 ```yaml

> # tasks file for Memcached

Reviewer: Why not provides the example direct? I mean, the final result with the config.

Author: My thought process was trying to provide a "before" and "after" so that users could follow along to what was actually changing.

Reviewer: I think is better if we just add an example with the final result.

- name: start memcached
  kubernetes.core.k8s:
    definition:
      ...
      spec:
        ...
        template:
          ...
          spec:
            affinity:
              nodeAffinity:
                requiredDuringSchedulingIgnoredDuringExecution:
                  nodeSelectorTerms:
                  - matchExpressions:
                    - key: kubernetes.io/arch
                      operator: In
                      values:
                      - amd64
                      - arm64
                      - ppc64le
                      - s390x
            containers:
              ...
 ```

#### Go Operators
In the `testdata/go/v3` samples app, you'll find several configuration objects that need a nodeAffinity specification in the `config` and `bundle/manifests` directories. The full list is as follows:
- `testdata/go/v3/memcached-operator/bundle/manifests/memcached-operator.clusterserviceversion.yaml`
- `testdata/go/v3/memcached-operator/config/default/manager_config_patch.yaml`
- `testdata/go/v3/memcached-operator/config/default/manager_webhook_patch.yaml`
- `testdata/go/v3/memcached-operator/config/default/manager_auth_proxy_patch.yaml`

> - `testdata/go/v3/memcached-operator/config/manager/manager.yaml`

Reviewer: All languages (with the exception of JAVA) scaffold the config directory.
I do not think that we should say in the testdata/ ... we can describe what authors should to do in the default scaffold of their project. We can link the testdata as an example. 

Also, not all these files should be updated to receive the nodeAffinity. 
Only `config/default/manager_config_patch.yaml` and `config/manager/manager.yaml`
The other ones are only patches. 

I think would be nice if you check the PR for update Kubebuilder: https://github.com/kubernetes-sigs/kubebuilder/pull/2906. Those changes will endup here as well, also look at the doc proposed there. I think we can shape this part here.

c/c @jmrodri

 Go operators may also create Pod-defining objects dynamically. An example of this kind of object would be in `testdata/go/v3/memcached-operator/controllers/memcached_controller.go`.
```go
Spec: corev1.PodSpec{
    SecurityContext: &corev1.PodSecurityContext{

>       ...

Reviewer: Why do we need this part? Why not just add "Following an example using Golang for you know how can you configure the nodeAffinity for the workloads in the reconciliation"

Author: So my thinking was:
1. Provide the general idea and the common ground between the operator types
2. Provide a before and after walkthrough to explain how those changes might look different for each of the operator types when those objects are defined in the different languages.

I could absolutely simplify the whole back-half of the doc down to - we leave the specifics of the transformation to include node-affinity in your `Pod` definitions as an exercise for the user.

However, I guess I was writing the document more for people like me. I don't know the specific syntactical differences between languages (and would struggle to even know where to look for such info), so I find it helpful to see an example where I can see exactly what changed so that I can replicate it each place where I can identify the pattern.

```yaml
kind: ClusterServiceVersion
metadata:
  annotations:
    alm-examples: |-
      [
        {
          "apiVersion": "cache.example.com/v1alpha1",
          "kind": "Memcached",
          "metadata": {
            "name": "memcached-sample"
          },
          "spec": {
            "AntiAffinity": "soft",
            "affinity": {
              "nodeAffinity": {
                "requiredDuringSchedulingIgnoredDuringExecution": {
                  "nodeSelectorTerms": [
                    {
                      "matchExpressions": [
                        {
                          "key": "kubernetes.io/arch",
                          "operator": "In",
                          "values": [
                              "amd64",
                              "arm64",
                              "ppc64le",
                              "s390x",
                          ]
                        },
                      ]
                    },
                  ]
                }
              }
            }
    ...
```

> Finally, if you're using the operator-sdk to deploy operator, it tooling would deploy `testdata/helm/memcached-operator/config/samples/cache_v1alpha1_memcached.yaml` via the reference in `testdata/helm/memcached-operator/config/samples/kustomization.yaml`.

Reviewer 2: ```suggestion
Finally, if you're using the operator-sdk to deploy the operator, its tooling would deploy `testdata/helm/memcached-operator/config/samples/cache_v1alpha1_memcached.yaml` via the reference in `testdata/helm/memcached-operator/config/samples/kustomization.yaml`.
```

> Since this file copies the defaults from the values.yaml, ensure that set nodeAffinity here too:

Reviewer 2: ```suggestion
Since this file copies the defaults from the `values.yaml`, ensure to set `nodeAffinity` here too:
```

```yamlapiVersion: cache.example.com/v1alpha1
kind: Memcached
metadata:
  name: memcached-sample
spec:
  # Default values copied from <project_dir>/helm-charts/memcached/values.yaml
  AntiAffinity: soft
  affinity:
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchExpressions:
          - key: kubernetes.io/arch
            operator: In
            values:
            - amd64
            - arm64
            - ppc64le
            - s390x
```

##### The StatefulSet Example

> The helm operator example has one last pod-defining object that doesn't declare a node affinity based on architecture. This is the StatefulSet defined in `testdata/helm/memcached-operator/helm-charts/memcached/templates/statefulset.yaml`. StatefulSets declare a `PodSpecTemplate` that will be rolled out onto each node in the cluster. In this case, podAntiAffinity rules are set to protect against pods of the same memcached instance from running on the same node.

Reviewer 2: ```suggestion
The helm operator example has one last pod-defining object that doesn't declare a node affinity based on architecture. This is the StatefulSet defined in `testdata/helm/memcached-operator/helm-charts/memcached/templates/statefulset.yaml`. StatefulSets declare a `PodSpecTemplate` that will be rolled out onto each node in the cluster. In this case, `podAntiAffinity` rules are set to protect against pods of the same memcached instance from running on the same node.
```

---
title: Heterogeneous Cluster Support
linkTitle: Heterogeneous Cluster Support
weight: 80
---
## Overview

> A heterogeneous cluster is a cluster in which the nodes differ from each other in a way that makes their containers incompatible with each other. Generally, containers are incomptabible if their platform specifcations don't match. Platform specifications generally include a operating system and CPU architecture. The rest of this document assumes that the user is developing an application that needs to be able to deploy cleanly in a cluster with nodes of mixed platforms. Because an operator deployed to a heterogeneous cluster may not support all of the platforms in the cluster, it is important to set up safeguards that prevent scheduling pods on incompatible nodes. These scheduling implications affect both the operator as well as its operands.

Reviewer 5: "multi-arch compute" now :)

---
title: Heterogeneous Cluster Support
linkTitle: Heterogeneous Cluster Support
weight: 80

> ---

Reviewer: We have the doc https://sdk.operatorframework.io/docs/advanced-topics/multi-arch/ so wdyt about we do the changes on top of this one instead of creating a new doc?

> 

Reviewer: ```suggestion
## Overview 
```

Author: Added an overview section.

> An operator deployed to a heterogeneous cluster may not support all of the node architectures in the cluster. This has scheduling implications for both the operator as well as its operands.

Reviewer: WDYT about we change it here for something like: 

This documentation aims to describe how you can develop Operators which can provide support for multiple platforms. It will help you out ensure that you provide a solution that can work on and can be distributed to be used for multiple platforms. In this way, that provides `"Heterogeneous Cluster Support"`.  

Overview
Any recommendations or best practices suggested by the Kubernetes community, such as how to [develop Operator pattern solutions](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/) or how to [use controller-runtime](https://github.com/kubernetes-sigs/controller-runtime) are good recommendations for those who are looking to build operator projects with operator-sdk. Also, see [Operator Best Practices](https://sdk.operatorframework.io/docs/best-practices/best-practices). However, here are some common recommendations.

Author: Yeah - so effectively, it would be good to define what heterogeneous cluster support actually means. I'll see if I can make that more clear in my second pass.

Author: I tried to incorporate a similar definition in my newest draft.

> 

Reviewer: WDYT about something like:

1) All workload images used in the Pods (for Operator and Operands) provide support for multiple platforms. To check what the platforms supported for each image used you can run `docker inspect <image`. 

Example:

2)  All Deployments/Pods (Operator and Operands) MUST have the `nodeAffinity` properly configured. Otherwise, ....

Example: (we can put here in k8s manifest syntax) 

a) Set the node affinity configuration for your Operator/manager by updating the config/manager file, i.e: 

b) Set the node addinity for all Deployment(s)/Pod(s) created by your Operator (Operands):

If your solution is Ansible or Helm language based you will need to ensure that all playbooks or helm charts are properly configuring the node affinity as in the above example. However, if your Operator is Golang based then, you will do that your code source as the following example:

Example for Golang language-based projects:

3) If your solution is integrated and will be distributed via OLM then: 

We need to add here the steps over what they must do to configure the CSV.
But if they update the config manifest and run make bundle all should be properly set.

Author: I tried to front-load the coverage for the basic nodeAffinity block, and provide a separate section covering the OLM CSV file.

In the operator sections of each type, I covered each of the changes that would need to be made in each file.

Finally, I added a section near the overview briefly talking about how to looking the architectures supported in a manifest list, as well as the operator-sdk validator for multi-arch. I also provided a forwarding link for the OLM guide.

> ### Setting Operator Default Architectures

Reviewer: I think we need to add the info about the CSV at the end.
Why?
I can have an Operator built with the tool and I should to the config for All Pods/Deployments. Therefore, integrating with OLM to publish in the catalogs is Optional. So, IMHO it would be better at the end with something like Now, if you integrate your project with OLM then ...

Author: Got it - so we want to treat the CSV bit as a special case of the overall story of setting podAffinity for your pods, not as separate step entirely (since it's ultimately optional).

Author: Separated this out into its own section.

### Setting Operand Default Architectures

The affinity block shown above can also be used to set the default architectures for operands. Because operands can be created and configured many different ways, this section highlights the most common places where resources will need updates. As a general rule, anywhere in the operator where a `PodSpec` or `PodTemplateSpec` is defined, you'll want to supply a corresponding affinity block covering the supported arches. This includes pod definitions as well as objects that create collections of pods - such as deployments, daemonsets, statefulsets, etc.

#### Go Operators
``
The most common directory to look at for objects that will need an architecture affinity block will is the operators ``config` directory. In the go samples app, located in the `testdata directory`, you'll find several Deployment configsthat need this updated including:
`testdata/go/v2/memcached-operator/config/default/manager_auth_proxy_patch.yaml`
`testdata/go/v2/memcached-operator/config/default/manager_webhook_patch.yaml`

> `testdata/go/v2/memcached-operator/config/manager/manager.yaml`

Reviewer: We should not use go/v2 that is deprecated and should no longer be used.

Author: Updated to v3

### Setting the Supported Architectures

Kubernetes provides a mechanism called `nodeAffinity` which can be used to limit the possible node targets where a pod can be scheduled. Here is an example of how to update a `PodSpec` or `PodTemplateSpec` to prevent the scheduling of pods on nodes of incompatible architectures. Here we compare the `kubernetes.io/arch` key to ensure that the value falls within on of the supported architectures. This assumes that the pod in question is referencing a manifest list which points to container images for each of the architectures listed. It is important to remember that `nodeAffinity` must be set anywhere a container reference is defined, including `Pod`, `Deployment`, `DaemonSet`, `StatefulSet`, or any other object that defines a `PodSpec` or `PodTemplateSpec`.

```yaml
affinity:
  nodeAffinity:

>     requiredDuringSchedulingIgnoredDuringExecution:

Reviewer: Why the requiredDuringSchedulingIgnoredDuringExecution is better than preferredDuringSchedulingIgnoredDuringExecution?

Author: Added a sentence after this block explicitly answering that question. :)

			Image:   "memcached:1.4.36-alpine",
			Name:    "memcached",
			Command: []string{"memcached", "-m=64", "-o", "modern", "-v"},
			Ports: []corev1.ContainerPort{{
			    ContainerPort: 11211,
				Name:          "memcached",
			}},
		}},
	},
},
 ```

Here, you would update the PodSpec by referencing the API directly, rather than supplying a yaml. The final result might look like this:
 ```go
	Spec: corev1.PodSpec{
        Affinity:
            NodeAffinity:
                RequiredDuringSchedulingIgnoredDuringExecution:
                    NodeSelectorTerms: []corev1.NodeSelectorTerm {{
                       MatchExpressions: []corev1.NodeSelectorRequirement {{
                           key: "kubernetes.io/arch"
                           operator: "In"
                           values: []string{"amd64", "arm64", "ppc64le", "s390x"}
                       }} 
                    }}
	Spec: corev1.PodSpec{
		Containers: []corev1.Container{{
			Image:   "memcached:1.4.36-alpine",
			Name:    "memcached",
			Command: []string{"memcached", "-m=64", "-o", "modern", "-v"},
			Ports: []corev1.ContainerPort{{
			    ContainerPort: 11211,
				Name:          "memcached",
			}},
		}},
	},

>  ```

Reviewer: this example is not accurate it would be like:

```go
	Spec: corev1.PodSpec{
					// TODO(user): If you project provide support to other platforms
					// then you must to change the NodeAffinity accordingly.
					// More info: https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#affinity-and-anti-affinity
					Affinity: &corev1.Affinity{
						NodeAffinity: &corev1.NodeAffinity{
							RequiredDuringSchedulingIgnoredDuringExecution: &corev1.NodeSelector{
								NodeSelectorTerms: []corev1.NodeSelectorTerm{
									{
										MatchExpressions: []corev1.NodeSelectorRequirement{
											{
												Key:      "kubernetes.io/arch",
												Operator: "In",
												Values:   []string{"amd64"},
											},
											{
												Key:      "kubernetes.io/os",
												Operator: "In",
												Values:   []string{"linux"},
											},
										},
									},
								},
							},
						},
					},
					SecurityContext: &corev1.PodSecurityContext{
```

Author: Thank you for providing this! :+1: 
I'm new to golang syntax, especially that which involves these API spec references. This was very helpful.
